### ISSUE TITLE
Innocent Scarlet Quail - Tokens minted during non-TDE cycles vote count is unfairly reduced and can't get max vote boost causing significant loss of yield

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2023-11-convergence-judging#176]
Innocent Scarlet Quail

high

# Tokens minted during non-TDE cycles vote count is unfairly reduced and can't get max vote boost causing significant loss of yield

## Summary

When locking tokens, the amount of time locked changes the voting power received from locking a token. This incentivizes users to lock for the max duration to get the most possible votes/rewards. This is problematic because locks are required to end on TDE cycles. The result is that users that stake on non-TDE cycles can never lock the max amount. These users are therefore disadvantaged and suffer loss of yield. If they wait to stake then they lose on yield while waiting and if they stake right away they will lose partial yield due to not being able to reach the max lock.

## Vulnerability Detail

[LockingPositionService.sol#L231-L254](https://github.com/sherlock-audit/2023-11-convergence/blob/main/sherlock-cvg/contracts/Locking/LockingPositionService.sol#L231-L254)

    function mintPosition(
        uint96 lockDuration,
        uint256 amount,
        uint64 ysPercentage,
        address receiver,
        bool isAddToManagedTokens
    ) external onlyWalletOrWhiteListedContract {
        require(amount > 0, "LTE");
        require(ysPercentage <= MAX_PERCENTAGE, "YS_%_OVER_100");
        require(ysPercentage % RANGE_PERCENTAGE == 0, "YS_%_10_MULTIPLE");
        require(lockDuration <= MAX_LOCK, "MAX_LOCK_96_CYCLES");

        ICvgControlTower _cvgControlTower = cvgControlTower;

        uint96 actualCycle = uint96(_cvgControlTower.cvgCycle());
        uint96 endLockCycle = actualCycle + lockDuration;

     **@audit locking duration is unfairly limited**
        require(endLockCycle % TDE_DURATION == 0, "END_MUST_BE_TDE_MULTIPLE");

When creating a lock it is required that the final cycle of the lock must be a TDE multiple. The fundamental problem with this is that veCVG/ysCVG received is as follows ([L584](https://github.com/sherlock-audit/2023-11-convergence/blob/main/sherlock-cvg/contracts/Locking/LockingPositionService.sol#L584)):

    (lockDuration * cvgLockAmount) / MAX_LOCK

It utilizes MAX_LOCK which is a constant 96 cycles. Meanwhile users are limited so that their end cycle is has to be a TDE multiple. This creates yield loss. Take the following example. The current cycle is 10. Any created lock MUST end on a TDE multiple and can't go over 96 weeks. The longest possible time the user can lock is only 86 cycles. If they attempted to lock for 96 then the final week would be 106 which isn't a multiple of 12. This means the user only receives a max:

    86 / 96 * cvgLockAmount = 89.6%

This 10.4% penalty is now applied to the user unfairly for the entire duration of their lock. This causes a definite and significant loss of yield to the user.

## Impact

Tokens minted during non-TDE cycles suffer unfair loss of yield.

## Code Snippet

[LockingPositionService.sol#L231-L312](https://github.com/sherlock-audit/2023-11-convergence/blob/main/sherlock-cvg/contracts/Locking/LockingPositionService.sol#L231-L312)

## Tool used

Manual Review

## Recommendation

I see two solutions to this:
    
    1) Tokens should be calculated on the real max (i.e. using 86 instead of 96 in the above scenario
    2) User should be allowed to lock for longer than 96 to the nearest even TDE (i.e. a 98 cycle lock in the above scenario)

**Comment by IAm0x52** [comment]
Escalate.

```When locking in a cycle before a TDE, a value called ysPartial is created, which is taken into account for that TDE. For the subsequent TDE, it's the ysTotal value that is considered. It's normal not to receive additional ysCvg & mgCvg when extending the time only.```

It appears the sponsor has misunderstood the submission. I did not mention anything at all in my submission to disagree with either of these statements.

```This serves as an incentive for the user to always lock for the maximum time.```

The exact point of my submission is that users are artificially PREVENTED from locking for the maximum time and therefore suffer loss of yield when locking on a non-TDE cycle.

As stated in my submission, the percentage of tokens received should be proportional to the TRUE max time that a user can lock. If a user can only lock for 88 cycles due to the limitation of artificially imposed then the token amount they receive should be in proportion to a max of 88 cycles not 96. I agree that the user should only get half for the first TDE since they only locked for 8/16 cycles but they should receive the max votes for the subsequent 80 cycles. Since they did in fact lock for the maximum time they were allowed to.

**Comment by sherlock-admin2** [comment]
> Escalate.
> 
> ```When locking in a cycle before a TDE, a value called ysPartial is created, which is taken into account for that TDE. For the subsequent TDE, it's the ysTotal value that is considered. It's normal not to receive additional ysCvg & mgCvg when extending the time only.```
> 
> It appears the sponsor has misunderstood the submission. I did not mention anything at all in my submission to disagree with either of these statements.
> 
> ```This serves as an incentive for the user to always lock for the maximum time.```
> 
> The exact point of my submission is that users are artificially PREVENTED from locking for the maximum time and therefore suffer loss of yield when locking on a non-TDE cycle.
> 
> As stated in my submission, the percentage of tokens received should be proportional to the TRUE max time that a user can lock. If a user can only lock for 88 cycles due to the limitation of artificially imposed then the token amount they receive should be in proportion to a max of 88 cycles not 96. I agree that the user should only get half for the first TDE since they only locked for 8/16 cycles but they should receive the max votes for the subsequent 80 cycles. Since they did in fact lock for the maximum time they were allowed to.

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by nevillehuang** [comment]
@walk-on-me @shalbe-cvg @walk-on-me 

I think @IAm0x52 is correct, you might want to revisit his comments and submissions, since it is crucial to the protocol logic to incentivize maximum lock time.

**Comment by Czar102** [comment]
I believe this is a valid protocol design. Per sponsor comment:
> As a lock has to finish on a TDE, if you are not on a TDE multiple you cannot lock max time but 95 cycles lets say for the cycle 1 or 13

A lock has to finish on TDE. Moreover, it would be more unfair to people who staked in earlier cycles modulo 12 if the number of cycles was rounded up, hence they stake for a longer time and don't receive additional rewards.

In the end, the timestamps at which the reward decreases is arbitrary per design. Also, the fact that the `MAX_LOCK` is unavailable at a large fraction of timestamps is not said to be important or enforced (to my best knowledge) and it seems it can be considered a "hypothetical" max lock duration.

Hence, planning to reject the escalation and leave the issue as is.

**Comment by IAm0x52** [comment]
My concern is the loss of yield caused by this arbitrary decision. The average user will not know to wait for this and once a lock has been created there is no way to ever recover this lost yield. Even if they wait then they earn no yield during that period and if they don't wait their yield on their deposit is permanently and artificially lowered. That is what in my opinion makes it medium but obviously the final decision is yours.

**Comment by Czar102** [comment]
> The average user will not know to wait for this

It seems there is an implied assumption of a misuse/misunderstanding of protocol mechanics in these considerations, which is not a sound argument for making the issue valid.

**Comment by IAm0x52** [comment]
That was only a side comment. The main argument is loss of yield. No matter the scenario if it is not a TDE cycle there is loss of yield for the depositor.

**Comment by Czar102** [comment]
But this loss of yield is with respect to a yield derived because of a misunderstanding of the actual mechanics of the protocol; at least that's how I understand it.

**Comment by IAm0x52** [comment]
What misunderstanding exactly? They are barred from being able to lock for the full duration and get proper yield because of incorrect protocol design.

**Comment by nevillehuang** [comment]
I agree with @IAm0x52, because I really see no reason why the protocol wouldn't want to incentivize more people to lock to participate in the protocol for voting/rewards (but instead let them lose out on yield explicitly), which will only benefit them. This should have been a design consideration/risk explicitly mentioned in the contest details, and since it is not, I believe it is a valid issue.

**Comment by Czar102** [comment]
Can't a user extend their lock at the nearest `cycle` divisible by `TDE_DURATION`? And the fact that their reward will be lower for their first TDE is expressing the fact that they didn't lock during the first cycle after a TDE.

@IAm0x52 @nevillehuang

**Comment by IAm0x52** [comment]
When you extend the lock it doesn't increase your votes, you're permanently locked to the lower vote count. The penalty for the partial TDE cycle is separate. Currently you are penalized twice. Once as you've pointed out by getting partial shares the partial cycle (which is expected) but additionally all of your subsequent TDE cycles are lower as well.

**Comment by Czar102** [comment]
Is extending the lock to a close date in the future maintaining high number of votes? This could allow for creation of short locks (they can be always unlocked soon) while the vote multiplier would stay high. The cost would be an ending maximal lock.

**Comment by IAm0x52** [comment]
Yeah your current vote count is extended indefinitely when you lock. So if you lock once for the max time, at the end of your max lock you could extend it one week at a time and keep your full vote count while always being max one week away from your lock ending. The opposite is true and if you lock for minimum time then extend for max duration, you would have the small amount for the entire max duration.

**Comment by Czar102** [comment]
Has this been raised to the protocol team? I think this would clearly be a bypass of the lock mechanism implied by the design.

**Comment by IAm0x52** [comment]
They are aware of it. Comments across the code as well indicate they are fully aware and intended it to not change when extending locks

**Comment by Czar102** [comment]
I thought about this issue a lot. Had some back-and-forth thoughts whether this finding should be valid or not.

**The documentation clearly implies that the code works as the sponsor intends.** I understand that from this design, odd properties follow (properties that seem to potentially be unwanted).

This is a consideration of the design and whether it is optimal. The fix is an improvement of the design. It is implied by the documentation that the `MAX_LOCK` lock time is not always available.

Hence, I will consider this issue low/informational.

**Comment by Czar102** [comment]
Result:
Low
Unique

**Comment by sherlock-admin2** [comment]
Escalations have been resolved successfully!

Escalation status:
- [IAm0x52](https://github.com/sherlock-audit/2023-11-convergence-judging/issues/176/#issuecomment-1869981328): rejected
